src/Http/WebUtilities/src/HttpResponsePipeWriter.cs

```diff
@@ -0,0 +1,275 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.IO;
+using System.IO.Pipelines;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.WebUtilities
+{
+    /// <summary>
+    /// Writes to the <see cref="PipeWriter"/> using the supplied <see cref="Encoding"/>.
+    /// It does not write the BOM and also does not close the stream.
+    /// </summary>
+    public class HttpResponsePipeWriter : TextWriter
+    {
+        private readonly Encoder _encoder;
+        private readonly char[] _singleCharArray;
+        private readonly PipeWriter _writer;
+
+        public override Encoding Encoding { get; }
+
+        private bool _disposed;
+
+        public HttpResponsePipeWriter(
+            PipeWriter writer,
+            Encoding encoding)
+        {
+            _writer = writer ?? throw new ArgumentNullException(nameof(writer));
+            Encoding = encoding ?? throw new ArgumentNullException(nameof(encoding));
+            _encoder = encoding.GetEncoder();
+            _singleCharArray = ArrayPool<char>.Shared.Rent(1);
+        }
+
+        public override void Write(char value)
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(nameof(HttpResponseStreamWriter));
+            }
+
+            _singleCharArray[0] = value;
+            var span = new Span<char>(_singleCharArray, 0, 1);
+            Write(span);
+        }
+
+        public override void Write(char[] values, int index, int count)
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(nameof(HttpResponseStreamWriter));
+            }
+
+            if (values == null || count == 0)
+            {
+                return;
+            }
+
+            var value = new Span<char>(values, index, count);
+            Write(value);
+        }
+
+        public override void Write(ReadOnlySpan<char> value)
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(nameof(HttpResponseStreamWriter));
+            }
+
+            if (value == null)
+            {
+                return;
+            }
+
+            var length = _encoder.GetByteCount(value, false);
+            var buffer = _writer.GetSpan(length);
+            _encoder.GetBytes(value, buffer, false);
```

Comment:
@GrabYourPitchforks - is there a more efficient way to do this? wondering about the pattern of `GetByteCount` immediately followed by `GetBytes`